clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -596,46 +614,51 @@ invariant_checker_t::print_results()
         // We do not bother to support legacy traces without encodings.
-        TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_)) {
-        addr_t trace_pc = shard->prev_instr_.instr.addr;
+        expect_encoding && type_is_instr_direct_branch(shard->prev_instr_.instr.type)) {
         if (shard->prev_instr_.instr.encoding_is_new)
-            shard->branch_target_cache.erase(trace_pc);
-        auto cached = shard->branch_target_cache.find(trace_pc);
+            shard->branch_target_cache.erase(shard->prev_instr_.instr.addr);
+        auto cached = shard->branch_target_cache.find(shard->prev_instr_.instr.addr);
         if (cached != shard->branch_target_cache.end()) {
             have_cond_branch_target = true;
             cond_branch_target = cached->second;
         } else {
-            instr_t instr;
-            instr_init(GLOBAL_DCONTEXT, &instr);
-            const app_pc decode_pc =
-                const_cast<app_pc>(shard->prev_instr_.instr.encoding);
-            const app_pc next_pc = decode_from_copy(
-                GLOBAL_DCONTEXT, decode_pc, reinterpret_cast<app_pc>(trace_pc), &instr);
-            if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
+            if (shard->prev_instr_decoded_ == nullptr ||
+                !opnd_is_pc(instr_get_target(shard->prev_instr_decoded_.get()))) {
                 // Neither condition should happen but they could on an invalid
                 // encoding from raw2trace or the reader so we report an
                 // invariant rather than asserting.
-                report_if_false(shard, false, "Branch target is not decodeable");
+                error_msg = "Branch target is not decodeable";
             } else {
                 have_cond_branch_target = true;
-                cond_branch_target =
-                    reinterpret_cast<addr_t>(opnd_get_pc(instr_get_target(&instr)));
-                shard->branch_target_cache[trace_pc] = cond_branch_target;
+                cond_branch_target = reinterpret_cast<addr_t>(
+                    opnd_get_pc(instr_get_target(shard->prev_instr_decoded_.get())));
+                shard->branch_target_cache[shard->prev_instr_.instr.addr] =
+                    cond_branch_target;
             }
-            instr_free(GLOBAL_DCONTEXT, &instr);
         }
     }
-
+    bool saw_repeated_syscall_instrs_with_same_pc = false;
+    if (cur_instr_decoded != nullptr && shard->prev_instr_decoded_ != nullptr &&
+        shard->prev_instr_.instr.addr != 0 /*first*/ &&
+        instr_is_syscall(cur_instr_decoded.get()) &&
+        memref.instr.addr == shard->prev_instr_.instr.addr &&
+        instr_is_syscall(shard->prev_instr_decoded_.get())) {
+        // Set this flag so that repeated syscalls are not
+        // double reporeted as other PC discontinuity below.
+        saw_repeated_syscall_instrs_with_same_pc = true;
+        error_msg = "Repeated syscall instrs with the same PC";
```

Comment:
It is confusing that we set the error_msg here but then go on to do the other checks. It'd be better if (similar to the branch target check) we set the error_msg at the end (inside the `if (!valid_nonbranch_flow) {` if block). Also see the other comment on how we're setting `valid_nonbranch_flow`.

Reply:
The other existing checks inside that block depend on ` if (shard->prev_instr_.instr.addr != 0 /*first*/) 
` which the Repeated syscall doesn't and the current test sequence only has instrs both of which are syscalls and the double syscall check would be skipped and results in a false negative. 

In an earlier PR I removed an instr before the first syscall in the test sequence after the comments. I can put that instr back infront of the first syscall so that I can put the double syscall check inside ` if (shard->prev_instr_.instr.addr != 0 /*first*/)`

Reply:
Actually nevermind. That check should still be fine and will move the double syscall check inside that if-block.  I will change the `error_msg` assignment for the `Branch target decode` check back to `report_if_false` and keep it where it is.